Update Azure API Guidelines for new LRO pattern (#353)
Microsoft/api-guidelines: azure/Guidelines.md

# Microsoft Azure REST API Guidelines

> <!-- cspell:ignore autorest, etag, idempotency, maxpagesize, innererror -->

Reviewer: Might be worth just putting a .vscode/cspell.json file in the repo.

GET    | Read the resource | `200-OK`
DELETE | Remove the resource | `204-No Content`\; avoid `404-Not Found`

> :white_check_mark: **DO** return status code `202-Accepted` and follow the guidance in [Long-Running Operations & Jobs](#long-running-operations--jobs) when a PUT, POST, or DELETE method completes asynchronously.

Reviewer 2: I think we should add subheadings to the LRO section to be more "what I'm trying to do" focused:
- Initiating an action via POST or deleting a resource: These both return 202s and a status monitor
- Creating a resource via PUT (or POST [note: I put POST in parens because we discourage POST for resource creation]): These both return 200/201 and the resource itself (not the status monitor)
- Polling an LRO
- Controlling an LRO
- Update an resource via POST-action; do not make LRO PATCH operations

It is OK to repeat some bullets in each section as devs will only read the section relevant to them. There has been so many problems with LROs that I think structuring the bullets this way will really help bring the improvements we seek.

need to change the structure of these LRO guidelines a tiny bit. I think the first part should describe the guidelines for POST/DELETE

Author: I've restructured roughly along the lines you suggest.  We need to settle on the pattern to use for POST to create a resource -- I currently have that grouped in the "standard LRO pattern" but can move it if necessary.

:white_check_mark: **DO** review any API changes with the Azure API Stewardship Board

> Clients specify the version of the API to be used in every request to the service, even requests to an `Operation-Location` or `nextLink` URL returned by the service.

Reviewer: Should this explanation go below the **DO** directive right below it? IIRC, I thought we put explanations after the requirement.

Author: I would be okay to move it, but I think this sentence "provides context" for the guidelines below rather than "explaining" them. That's why I put it above.

Reviewer: Okay. Just a nit. Works for me either way. :)

:no_entry: **DO NOT** use a long-running POST to create a resource -- use PUT as described below.

:white_check_mark: **DO** allow the client to pass an `Operation-Id` header with an ID for the operation's status monitor.

:white_check_mark: **DO** generate an ID (typically a GUID) for the status monitor if the `Operation-Id` header was not passed by the client.

:white_check_mark: **DO** fail a request with a `400-BadRequest` if the `Operation-Id` header matches an outstanding operation unless the request is identical to the prior request (a retry scenario).

:white_check_mark: **DO** perform as much validation as practical when initiating the operation to alert clients of errors early.

:white_check_mark: **DO** return a `202-Accepted` status code from the request that initiates an LRO if the processing of the operation was successfully initiated (except for "PUT with additional processing" type LRO).

:warning: **YOU SHOULD NOT** return any other `2xx` status code from the initial request of an LRO -- return `202-Accepted` and a status monitor even if processing was completed before the initiating request returns.

> :ballot_box_with_check: **YOU SHOULD** include an `Operation-Location` header in the response with the absolute URL of the status monitor for the operation, but do not include an api-version query parameter.

Reviewer 3: How does a client know it is supposed to add back the api-version?

> :white_check_mark: **DO** return a status monitor in the response body as described in [Status and results of long-running operations](#status-and-results-of-long-running-operations).

Reviewer: Looks like the heading changed to "Obtaining status and results of long-running operations" so the bookmark needs to change as well. To keep these stable, you can use A.name in the heading and just reference that everywhere:

```markdown
See [here](#there).

## Some heading <a name="there"></a>
```

#### PUT operation with additional long-running processing

For a PUT (create or replace) with additional long-running processing:

:white_check_mark: **DO** allow the client to pass an `Operation-Id` header with a ID for the status monitor for the operation.

:white_check_mark: **DO** generate an ID (typically a GUID) for the status monitor if the `Operation-Id` header was not passed by the client.

> :white_check_mark: **DO** fail a request with a `400-BadRequest` if the `Operation-Id` header that matches an outstanding operation unless the request is identical to the prior request (a retry scenario).

Reviewer 3: What happens if the operation completes before the retry is processed?  Is there an expectation that the service keep tombstones of operation-ids?

Author: Yes.  I will clarify that. "outstanding" is the wrong word here.

Author: I think the right word is "existing". We recommend that the operation status resource be retained for at least 24 hours before it is garbage collected. This gives a guarantee of idempotency for at least 24 hours, which should be sufficient for the vast majority of cases.

operation is being processed. The client obtains the outcome of the operation at some later time
through another API call.
See the [Long Running Operations section](./ConsiderationsForServiceDesign.md#long-running-operations) in
Considerations for Service Design for an introduction to the design of long-running operations.

:white_check_mark: **DO** implement an operation as an LRO if the 99th percentile response time is greater than 1s.

> :no_entry: **DO NOT** implement PATCH as an LRO.  If LRO update is required it should be implemented with POST.

Reviewer 4: ```suggestion
:no_entry: **DO NOT** implement PATCH as an LRO.  If LRO update is required it must be implemented with POST.
```
Let's have the language match--DO NOT = MUST (as compared to the weaker 'should')

In rare instances where an operation may take a _very long_ time to complete, e.g. longer than 15 minutes,
it may be better to expose this as a first class resource of the API rather than as an operation on another resource.

> There are two basic patterns for long-running operations in Azure. The first pattern is used for a POST and DELETE

Reviewer 4: It would be valuable if we could agree on a name for these two patterns. I'm thinking of the [Gang of Four](https://www.amazon.com/Design-Patterns-Elements-Reusable-Object-Oriented/dp/0201633612/ref=sr_1_1?keywords=9780321700698&linkCode=qs&qid=1654610601&s=books&sr=1-1) where they name things like, `flywheel`, `state` etc. Although it sounds simple, I think this will help us start to develop a vocabulary in our API reviews.

operations that initiate the LRO. These return a `202 Accepted` response with a JSON status monitor in the response body.

> The second pattern applies only in the case of a PUT operation to create a resource that also involves additional long-running processing.

Reviewer 4: ```suggestion
The second pattern applies only in the case of a PUT operation to create a resource that also involves additional long-running processing. For guidance on when to use a specific pattern, please refer to [Considerations for Service Design, Long Running Operations](./ConsiderationsForServiceDesign.md#long-running-operations)
```

These are described in the following two sections.

#### POST or DELETE LRO pattern

A POST or DELETE long-running operation accepts a request from the client to initiate the operation processing and returns
a [status monitor](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.3) that reports the operation's progress.

:no_entry: **DO NOT** use a long-running POST to create a resource -- use PUT as described below.

:white_check_mark: **DO** allow the client to pass an `Operation-Id` header with an ID for the operation's status monitor.

:white_check_mark: **DO** generate an ID (typically a GUID) for the status monitor if the `Operation-Id` header was not passed by the client.

> :white_check_mark: **DO** fail a request with a `400-BadRequest` if the `Operation-Id` header that matches an outstanding operation unless the request is identical to the prior request (a retry scenario).

Reviewer: Wording is off here. Not sure what you were going for. Perhaps get rid of "that" in "that matches"?

:white_check_mark: **DO** perform as much validation as practical when initiating the operation to alert clients of errors early.

> :white_check_mark: **DO** return a `202-Accepted` status code from the request that initiates an LRO if the processing of the operation was successfully initiated (except for "create (PUT) with additional processing" type LRO).

Reviewer: Would you want to change this to just "PUT with additional processing" as well?

:warning: **YOU SHOULD NOT** return any other `2xx` status code from the initial request of an LRO -- return `202-Accepted` and a status monitor even if processing was completed before the initiating request returns.

:white_check_mark: **DO** include a `Retry-After` header in the response to the initiating request if the operation is not complete. The value of this header should be an integer number of seconds to wait before making the first request to the status monitor.

:ballot_box_with_check: **YOU SHOULD** include an `Operation-Location` header in the response with the absolute URL of the status monitor for the operation, but do not include an api-version query parameter.

:white_check_mark: **DO** return a status monitor in the response body as described in [Status and results of long-running operations](#status-and-results-of-long-running-operations).

> #### Create (PUT) operation with additional long-running processing

Author: ```suggestion
#### PUT operation with additional long-running processing
```

> For a create (PUT) with additional long-running processing:

Author: ```suggestion
For a create or replace (PUT) with additional long-running processing:
```

:white_check_mark: **DO** allow the client to pass an `Operation-Id` header with a ID for the status monitor for the operation.

:white_check_mark: **DO** generate an ID (typically a GUID) for the status monitor if the `Operation-Id` header was not passed by the client.

:white_check_mark: **DO** fail a request with a `400-BadRequest` if the `Operation-Id` header that matches an outstanding operation unless the request is identical to the prior request (a retry scenario).

:white_check_mark: **DO** perform as much validation as practical when initiating the operation to alert clients of errors early.

> :white_check_mark: **DO** return a `201-Created` status code from the initial request with a representation of the resource if the resource was created successfully.

Author: ```suggestion
:white_check_mark: **DO** return a `201-Created` status code, or `200-OK` for replace, from the initial request with a representation of the resource if the resource was created or replaced successfully.
```

> :white_check_mark: **DO** include an `Operation-Id` header in the response with the ID of the status monitor for the operation.

Reviewer 5: Many of these are the same.  Would it be more helpful to aggregate and only call out the differences based on verb (i.e., how much do you want to bet someone will miss the 201 vs. 202 because the first few guidelines were identical so their brain short circuited the rest)?

:white_check_mark: **DO** include a `Retry-After` header in the response to the initiating request if the operation is not complete. The value of this header should be an integer number of seconds to wait before making the first request to the status monitor.

:ballot_box_with_check: **YOU SHOULD** include an `Operation-Location` header in the response with the absolute URL of the status monitor for the operation, but do not include an api-version query parameter.

#### Obtaining status and results of long-running operations

For all long-running operations, the client will issue a GET on a status monitor resource to obtain the current status of the operation.

:white_check_mark: **DO** support the GET method on the status monitor endpoint that returns a `200-OK` response with the current state of the status monitor.

:white_check_mark: **DO** return a status monitor in the response body that conforms with the following structure:

**OperationStatus** : Object

Property | Type        | Required | Description
-------- | ----------- | :------: | -----------
`id`     | string      | true     | The unique id of the operation

> other    |             | true     | Other values needed for a GET request to the status monitor (e.g. location)

Reviewer 4: Wondering if we should be more clear about this, e.g. if it should be an object, e.g. `additionalProperties`. Does this help with breaking changes, e.g. if I need to add something in the future?

Reviewer 5: Does it mean literally `additionalProperties: true`?  Or do we mean extra well defined values?  I'd assume extra well defined values?

`status` | string      | true     | enum that includes terminal values "Succeeded", "Failed", "Canceled"
`error`  | ErrorDetail |          | Error object that describes the error when status is "Failed"

> `result` | object      |          | Only for POST action-type LRO, the results of the operation when completed successfully

Reviewer 5: How does this work for partial success in batched APIs?  Should we call out that those all go in the `result` and `error` is only for entire batch failure?

Author: I think we should cover this when we add guidance on batch operations.

:white_check_mark: **DO** include a `Retry-After` header in the response to GET requests to the status monitor if the operation is not complete. The value of this header should be an integer number of seconds to wait before making the next request to the status monitor.

:white_check_mark: **DO** include the `result` property (if any) in the status monitor for a POST action-type long-running operation when the operation completes successfully.

:no_entry: **DO NOT** include a `result` property in the status monitor for a long-running operation that is not a POST action-type long-running operation.

> :white_check_mark: **DO** retain the status monitor resource for some documented period of time (at least 24 hours) after the operation completes.

Reviewer 5: ```suggestion
:white_check_mark: **DO** retain the status monitor resource for some publicly documented period of time (at least 24 hours) after the operation completes.
```

operation is being processed. The client obtains the outcome of the operation at some later time
through another API call.
See the [Long Running Operations section](./ConsiderationsForServiceDesign.md#long-running-operations) in
Considerations for Service Design for an introduction to the design of long-running operations.

:white_check_mark: **DO** implement an operation as an LRO if the 99th percentile response time is greater than 1s.

:no_entry: **DO NOT** implement PATCH as an LRO.  If LRO update is required it should be implemented with POST.

In rare instances where an operation may take a _very long_ time to complete, e.g. longer than 15 minutes,
it may be better to expose this as a first class resource of the API rather than as an operation on another resource.

> There are two basic patterns for long-running operations in Azure. In the first pattern is used for a POST or DELETE

Reviewer 2: ```suggestion
There are two basic patterns for long-running operations in Azure. The first pattern is used for a POST and DELETE
```

> operation that initiates the LRO returns a `202 Accepted` response with a status monitor in the response body.

Reviewer 2: ```suggestion
operations that initiate the LRO. These return a `202 Accepted` response with a JSON status monitor in the response body.
```

> The second pattern is applies only in the case of a PUT operation to create a resource that also involves additional long-running processing.

Reviewer 2: ```suggestion
The second pattern applies only in the case of a PUT operation to create a resource that also involves additional long-running processing.
```

These are described in the following two sections.

#### POST or DELETE LRO pattern

A POST or DELETE long-running operation accepts a request from the client to initiate the operation processing and returns

> a [status monitor](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.3) that will report the operation results.

Reviewer 2: ```suggestion
a [status monitor](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.3) that reports the operation's progress.
```

> :white_check_mark: **DO** allow the client to pass an `Operation-Id` header with a ID for the status monitor for the operation.

Reviewer 2: ```suggestion
:white_check_mark: **DO** allow the client to pass an `Operation-Id` header with an ID for the operation's status monitor.
```

> :white_check_mark: **DO** generate an ID (typically a GUID) of the status monitor if the `Operation-Id` header was not passed by the client.

Reviewer 2: ```suggestion
:white_check_mark: **DO** generate an ID (typically a GUID) for the status monitor if the `Operation-Id` header was not passed by the client.
```

:warning: **YOU SHOULD NOT** return any other `2xx` status code from the initial request of an LRO -- return `202-Accepted` and a status monitor even if processing was completed before the initiating request returns.

:white_check_mark: **DO** include a `Retry-After` header in the response to the initiating request if the operation is not complete. The value of this header should be an integer number of seconds to wait before making the first request to the status monitor.

:ballot_box_with_check: **YOU SHOULD** include an `Operation-Location` header in the response with the absolute URL of the status monitor for the operation, but do not include an api-version query parameter.

:white_check_mark: **DO** return a status monitor in the response body as described in [Status and results of long-running operations](#status-and-results-of-long-running-operations).

#### Create (PUT) operation with additional long-running processing

For a create (PUT) with additional long-running processing:

:white_check_mark: **DO** allow the client to pass an `Operation-Id` header with a ID for the status monitor for the operation.

> :white_check_mark: **DO** generate an ID (typically a GUID) of the status monitor if the `Operation-Id` header was not passed by the client.

Reviewer 2: ```suggestion
:white_check_mark: **DO** generate an ID (typically a GUID) for the status monitor if the `Operation-Id` header was not passed by the client.
```
